overlord/devicestate/devicemgr.go

```diff
@@ -995,23 +995,62 @@ func (m *DeviceManager) Systems() ([]*System, error) {
 
 var ErrUnsupportedAction = errors.New("unsupported action")
 
+// Reboot triggers a reboot into the given systemLabel and mode.
+//
+// When called without a systemLabel and without a mode it will just
+// trigger a regular reboot.
+//
+// When called without a systemLabel but with a mode it will use
+// the current system to enter the given mode.
+//
+// Note that "recover" and "run" modes are only available for the
+// current system.
+func (m *DeviceManager) Reboot(systemLabel, mode string) error {
+	// most simple case: just reboot
+	if systemLabel == "" && mode == "" {
+		m.state.Lock()
+		defer m.state.Unlock()
+
+		logger.Noticef("restarting system")
+		m.state.RequestRestart(state.RestartSystemNow)
+		return nil
+	}
+
+	// no systemLabel means "current" so get the current system label
+	if systemLabel == "" {
+		systemMode := m.SystemMode()
+		currentSys, err := currentSystemForMode(m.state, systemMode)
+		if err != nil {
+			return fmt.Errorf("cannot get curent system: %v", err)
+		}
+		systemLabel = currentSys.System
+	}
+
+	return m.doSystemAction(systemLabel, SystemAction{Mode: mode})
+}
+
 // RequestSystemAction request provided system to be run in a given mode. A
 // system reboot will be requested when the request can be successfully carried
 // out.
 func (m *DeviceManager) RequestSystemAction(systemLabel string, action SystemAction) error {
 	if systemLabel == "" {
 		return fmt.Errorf("internal error: system label is unset")
 	}
+	return m.doSystemAction(systemLabel, action)
+}
 
+func (m *DeviceManager) doSystemAction(systemLabel string, action SystemAction) error {
 	if err := checkSystemRequestConflict(m.state, systemLabel); err != nil {
 		return err
 	}
 
 	systemMode := m.SystemMode()
+	// XXX: why do we ignore the error here?
 	currentSys, _ := currentSystemForMode(m.state, systemMode)
 
 	systemSeedDir := filepath.Join(dirs.SnapSeedDir, "systems", systemLabel)
 	if _, err := os.Stat(systemSeedDir); err != nil {
+		// XXX: should we wrap this instead return a naked stat error?
```

Comment:
seems a not bad idea

Reply:
The caller in daemon calls os.IsNotExist() on this error, maybe the easies would be do have a separate ErrUnknonwnSystem or somesuch here to make it explicit and just wrap any other errors.

Reply:
Yeah, I think I will also do this in a followup to make this not bigger than it already is.